third_party/move/move-vm/runtime/src/loader/single_signature_loader.rs

```diff
@@ -0,0 +1,199 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use super::type_loader::intern_type;
+use move_binary_format::{
+    access::ModuleAccess,
+    binary_views::BinaryIndexedView,
+    errors::{PartialVMError, PartialVMResult},
+    file_format::{Bytecode, CodeUnit, CompiledModule, CompiledScript, SignatureIndex},
+};
+use move_core_types::vm_status::StatusCode;
+use move_vm_types::loaded_data::{runtime_types::Type, struct_name_indexing::StructNameIndex};
+use std::collections::BTreeMap;
+
+/// Maps indices of single signatures to their types. For example, vector bytecode instruction
+/// carries an index of a signature of vector element. This loader ensures that all these indices
+/// map to their runtime types.
+struct SingleSignatureMap<'a> {
+    /// Script or a module.
+    view: BinaryIndexedView<'a>,
+    /// The map from single signature index to its runtime type.
+    index_to_type: BTreeMap<SignatureIndex, Type>,
+    /// Interned struct names as indices. The caller is repsponsible for ensuring that all names
+    /// are interned and included here.
+    struct_idxs: &'a [StructNameIndex],
+}
+
+/// Creates a map of all signle signature indices to their types for a script.
+///
+/// The caller is repsponsible for ensuring that all struct indices (interned names) are included
+/// here.
+
+pub(crate) fn load_single_signatures_for_script(
+    script: &CompiledScript,
+    struct_idxs: &[StructNameIndex],
+) -> PartialVMResult<BTreeMap<SignatureIndex, Type>> {
+    let mut map = SingleSignatureMap {
+        view: BinaryIndexedView::Script(script),
+        index_to_type: BTreeMap::new(),
+        struct_idxs,
+    };
+
+    map.load_single_signatures_for_code_unit(&script.code)?;
+
+    Ok(map.index_to_type)
+}
+
+/// Creates a map of all signle signature indices to their types for a module.
+///
+/// The caller is repsponsible for ensuring that all struct indices (interned names) are included
+/// here.
+pub(crate) fn load_single_signatures_for_module(
+    module: &CompiledModule,
+    struct_idxs: &[StructNameIndex],
+) -> PartialVMResult<BTreeMap<SignatureIndex, Type>> {
+    let mut map = SingleSignatureMap {
+        view: BinaryIndexedView::Module(module),
+        index_to_type: BTreeMap::new(),
+        struct_idxs,
+    };
+
+    for function_def in module.function_defs() {
+        if let Some(code_unit) = &function_def.code {
+            map.load_single_signatures_for_code_unit(code_unit)?;
+        }
+    }
+
+    Ok(map.index_to_type)
+}
+
+impl<'a> SingleSignatureMap<'a> {
+    fn load_single_signatures_for_code_unit(&mut self, code: &CodeUnit) -> PartialVMResult<()> {
+        for instr in &code.code {
+            use Bytecode::*;
+            match instr {
+                // Instructions with single signature index.
+                VecPack(idx, _)
```

Comment:
Would it be instead useful to have a `get_single_signature_index() -> Option<SignatureIndex>` method on `Bytecode`? Or do you think this is a one off?

Either way, it is great to have explicit enum coverage!

Reply:
Why not, added `get_single_signature_index` 👍